auto_round/utils.py

```diff
@@ -664,3 +664,161 @@ def dynamic_import_inference_linear(backend, bits, group_size, sym):
     else:
         from auto_round_extension.cuda.qliner_triton import QuantLinear
     return QuantLinear
+
+
+import numba
+class Pakcer:
+    @staticmethod
+    def pack_2d_tensor(
+        raw_tensor: torch.Tensor, n_pack: int, bits: int, compression_dtype: torch.dtype = torch.int32
+    ) -> torch.Tensor:
+        """Packs a 2D tensor into a compressed tensor.
+        
+        This is a refer implementation for verifying the correctness of optimized implementation.
+
+        Args:
+            raw_tensor (torch.Tensor): The input tensor to be packed. The shape of the tensor should be (out_feats, in_feats) or (1, in_feats).
+            n_pack (int): The number of elements to pack together.
+            bits (int): The number of bits to use for each packed element.
+            compression_dtype (torch.dtype, optional): The data type of the compressed tensor. Defaults to torch.int32.
+
+        Returns:
+            torch.Tensor: The packed tensor.
+
+        """
+        raw_array = raw_tensor.cpu().numpy()
+        target_len = np.ceil(raw_array.shape[1] / n_pack).astype(int)
+        target_dtype = torch.tensor(0, dtype=compression_dtype).numpy().dtype
+        packed_array = np.zeros((raw_array.shape[0], target_len), dtype=target_dtype)
+        mask = np.uint8(2**bits - 1)
+        for j in range(packed_array.shape[1]):
+            start = n_pack * j
+            end = n_pack * (j + 1)
+            tmp = raw_array[:, start:end].astype(target_dtype)
+            tmp &= mask
+            for e in range(tmp.shape[1]):
+                tmp[:, e] = np.left_shift(tmp[:, e], bits * e)
+                packed_array[:, j] |= tmp[:, e]
+        packed_tensor = torch.from_numpy(packed_array).to(device=raw_tensor.device)
+        return packed_tensor
+
+
+    @staticmethod
+    @numba.jit(nopython=True, parallel=True, cache=True)
+    def pack_tensor_with_numpy_opt_np_numba(
+        raw_tensor: np.ndarray, n_pack: int, bits: int, compression_dtype=np.int32
+    ) -> np.ndarray:
+        out_features, in_features = raw_tensor.shape
+        new_in_features = (in_features + n_pack - 1) // n_pack
+        packed_tensor = np.zeros((out_features, new_in_features), dtype=compression_dtype)
+        raw_tensor = raw_tensor.astype(compression_dtype)
+
+        if bits == 4:
+            for i in range(new_in_features):
+                packed_tensor[:, i] = (
+                    (raw_tensor[:, i * n_pack + 7] << 28)
+                    | (raw_tensor[:, i * n_pack + 6] << 24)
+                    | (raw_tensor[:, i * n_pack + 5] << 20)
```

Comment:
is the packing logic same with autogptq exllama? better integrate it into one format and have a accuracy check